packages/block-editor/src/store/selectors.js

```diff
@@ -201,10 +217,11 @@ export const getBlocks = createSelector(
 			getBlock( state, clientId )
 		);
 	},
-	( state ) => [
-		state.blocks.byClientId,
-		state.blocks.order,
-		state.blocks.attributes,
+	( state, rootClientId ) => [
+		...map(
+			state.blocks.order[ rootClientId || '' ],
+			( id ) => state.blocks.cache[ id ]
+		),
 	]
```

Comment:
Is the shallow clone / concatenation necessary?

```suggestion
( state, rootClientId ) => map(
	state.blocks.order[ rootClientId || '' ],
	( id ) => state.blocks.cache[ id ]
)
```

Generally, it's kinda a misuse of `rememo`'s `getDependants`, since the array is highly unstable.

Also worth considering if the cost of `map`, combined with the cost of cache lookup, is actually worth the potential benefit. I suppose it's pretty important that `getBlocks` would return a consistent array reference each call, which would not happen without memoization by nature of the fact that the implementation's `map` would always return a new function.

Reply:
it's not a  problem that map always returns a new array, what's important is that the element of the arrays don't change (all getDependents functions return new arrays on each call)

Reply:
>it's not a problem that map always returns a new array, what's important is that the element of the arrays don't change (all getDependents functions return new arrays on each call)

I meant the _other_ map, the one in the implementation of the selector itself:

https://github.com/WordPress/gutenberg/blob/058c23a54c3bdc39b7c907b013b4f3ccf56bd972/packages/block-editor/src/store/selectors.js#L216

Reply:
> I suppose it's pretty important that getBlocks would return a consistent array reference each call,

This is the important  bit here and not the performance. If the value is different, onChange calls will be triggered  creating dirtiness. How would you retrieve the same instance when blocks don't change otherwise?

Reply:
> Also worth considering if the cost of map, combined with the cost of cache lookup, is actually worth the potential benefit

I didn't notice any adverse affects running the performance test. The main benefit is functional: we can detect that a change has occurred in the block list that's only related to the blocks that we control just by comparing the old and new references of the selector return value. This works because the cache reducer no longer traverses up the block tree past an "inner block controller". So the cache key of a given block won't change if a deeply nested block changes which is controlled by something else.
 
> Is the shallow clone / concatenation necessary?

The goal is for the selector to only change when a cache key of a direct descendant of the given root changes.

I'm definitely happy to improve this if you have ideas :) One thought is adding another state cache key for getBlocks

Reply:
>The goal is for the selector to only change when a cache key of a direct descendant of the given root changes.

What I mean is that both of these implementations will result in the same array members:

```js
const x = [ ...map( foo, fn ) ]
```
```js
const x = map( foo, fn )
```

The difference in practice is that the former (similar to what's implemented here) could incur an additional cost associated with the spread operation, in addition to simply being more code.

Reply:
Gotcha, that makes sense. I can put up something for this change. does `createSelector` care if the reference of the array changes?

Reply:
https://github.com/WordPress/gutenberg/pull/22366 fixes the issue and adds a test to verify behavior

Reply:
> Gotcha, that makes sense. I can put up something for this change. does `createSelector` care if the reference of the array changes?

It shouldn't make a difference.

And in practice, both implementations are going to create a new array every time.

Reply:
👍 that makes sense.